xnmt/lstm.py

```diff
@@ -9,28 +9,81 @@
 from xnmt.transducer import SeqTransducer, FinalTransducerState
 from xnmt.serialize.tree_tools import Ref, Path
 
+class UniLSTMState(object):
+  """
+  State object for UniLSTMSeqTransducer.
+  """
+  def __init__(self, network, prev=None, c=None, h=None):
+    self._network = network
+    if c is None:
+      c = [dy.zeroes(dim=(network.hidden_dim,)) for _ in range(network.num_layers)]
+    if h is None:
+      h = [dy.zeroes(dim=(network.hidden_dim,)) for _ in range(network.num_layers)]
+    self._c = tuple(c)
+    self._h = tuple(h)
+    self._prev = prev
+
+  def add_input(self, x):
+    new_c, new_h = self._network.add_input_to_prev(self, x)
+    return UniLSTMState(self._network, prev=self, c=new_c, h=new_h)
+
+  def b(self):
+    return self._network
+
+  def h(self):
+    return self._h
+
+  def s(self):
+    return self._c + self._h
+
+  def prev(self):
+    return self._prev
+
+  def set_h(self, es=None):
+    if es is not None:
+      assert len(es) == self._network.num_layers
+    self._h = tuple(es)
+    return self
+
+  def set_s(self, es=None):
+    if es is not None:
+      assert len(es) == 2 * self._network.num_layers
+    self._c = tuple(es[:self._network.num_layers])
+    self._h = tuple(es[self._network.num_layers:])
+    return self
+
+  def output(self):
+    return self._h[-1]
+
+
 class UniLSTMSeqTransducer(SeqTransducer, Serializable):
   """
   This implements a single LSTM layer based on the memory-friendly dedicated DyNet nodes.
   It works similar to DyNet's CompactVanillaLSTMBuilder, but in addition supports
   taking multiple inputs that are concatenated on-the-fly.
-  
-  Currently only supports transducing a complete sequence at once.
-  
+
   Args:
     exp_global (ExpGlobal): ExpGlobal object to acquire DyNet params and global settings. By default, references the experiment's top level exp_global object.
+    layers (int): number of layers
     input_dim (int): input dimension; if None, use exp_global.default_layer_dim
     hidden_dim (int): hidden dimension; if None, use exp_global.default_layer_dim
     dropout (float): dropout probability; if None, use exp_global.dropout
     weightnoise_std (float): weight noise standard deviation; if None, use exp_global.weightnoise_std
     param_init (ParamInitializer): how to initialize weight matrices; if None, use ``exp_global.param_init``
     bias_init (ParamInitializer): how to initialize bias vectors; if None, use ``exp_global.bias_init``
   """
-  def __init__(self, exp_global=Ref(Path("exp_global")), input_dim=None, hidden_dim=None,
-               dropout = None, weightnoise_std=None, param_init=None, bias_init=None):
+  yaml_tag = '!UniLSTMSeqTransducer'
+
+  def __init__(self, exp_global=Ref(Path("exp_global")), layers=1, input_dim=None, hidden_dim=None,
+               dropout=None, weightnoise_std=None, param_init=None, bias_init=None,
+               yaml_path=None, decoder_input_dim=None, decoder_input_feeding=True):
```

Comment:
The new arguments should be documented.

Reply:
`yaml_path` as well, even though it's kind of for internal use? Because the `decoder_...` arguments only get interpreted when `"decoder" in yaml_path` ...

Reply:
Hm, maybe you could add ```yaml_path``` as an argument but leave the description empty? And then somehow the information should be conveyed that ```decoder_...``` is only interpreted when ```yaml_path``` is given.